### PR TITLE
Fix web display of merge trains

### LINKED ISSUE #162: Order PRs in the web interface by approval
In the web interface, PRs are currently ordered by their ID:

* `PR#1` :yellow_circle: 
* `PR#2` :yellow_circle: 
* `PR#3` :heavy_check_mark: 

Supposing we get `merge` commands in the order 2, 3 and 1, we should display PRs in "train order" as:

* `PR#2` :yellow_circle: 
* `PR#3` :heavy_check_mark:  
* `PR#1` :yellow_circle: 

This only became a problem with the introduction of merge trains in v0.27.0 (#77, #137).  Before we could have only one PR building per project, so there was only one possible order.  Now, since we can have multiple PRs, we should display them in train order.

### LINKED ISSUE #163: Display speculative rebase failures properly on the web interface
The web interface fails to display PRs with speculative rebase failures.  They simply are not shown.  We should display them properly.

## Now there should be five PRs, but only four appear:

There was a speculative rebase failure:

### PR DESCRIPTION
**Pull request by rudymatela**
Closes: #162 
Closes: #163 

This PR fixes the web display of merge trains:

- [x] display PRs in the order they were approved which is the order of the train
    (closes #162);
- [x] display speculative rebase failures
    (closes #163);
- [x] refactor `WebInterface` to remove repeated code with the `classifiedPullRequests` function;
- [x] different icons and text for all possible integration statuses: `🟡❌✅ 🔷❗`;
- [x] add some automated tests to `classifiedPullRequests`;
- [x] take screenshots and place them here;

------

## Before merge actions (unchanged)

0-before

:arrow_up: Nothing changes when all PRs are waiting for review.

These PRs were approved in the following order:

* \#141
* \#143
* \#142
* \#144
* \#140

------

## Before (repository view)

before-repo

:arrow_up: note PR \#142 (Squidward) is missing and PRs are not listed in order of approval.

------

## After (repository view)

repo-after

:arrow_up: correct order, no duplicates, icons

------

## Before (global view)

before-global

:arrow_up: note PR \#142 (Squidward) is missing and PRs are not listed in order of approval.  There is one PR that is duplicate!

------

## After (global view)

global-after

:arrow_up: correct order, no duplicates, icons

------

### CODE REVIEW

**Comment by rudymatela** [comment]
There were changes on master.  Better rebase now than later...

**Comment by rudymatela** [comment]
This is ready for review.

@frank-channable or @alex-mckenna can either of you please review?  (whoever is free first -- first-come-first-merged)

File: src/WebInterface.hs [hunk 1]
```diff
@@ -243,7 +266,10 @@ viewPullRequest info pullRequestId pullRequest = do
         (BuildStarted ciUrl)       -> span " | " >> ciLink ciUrl "CI build"
         (BuildFailed (Just ciUrl)) -> span " | " >> ciLink ciUrl "CI build"
         _                          -> pure ()
-    _ -> pure ()
+    Conflicted _ _      -> span "  | " >> span "‼️ conflicted"
```
Comment (alex-mckenna) [inline src/WebInterface.hs:269]: Should this !! be some `:github-emoji-thing:` instead for consistency?

**Review by alex-mckenna — COMMENTED** [review]
I have no real comments about the code, it all looks pretty clean to me. I do wonder if now that more is displayed it's losing some clarity.

The emoji indicators are cute, but as we have more they become less obvious. The blue diamond for promoted feels odd, but I can't think of a good choice for this. Text only also has problems though, because it becomes a bit more difficult to just scan the page with your eyes... What do you think?

File: src/WebInterface.hs (hunk 1 above)
Comment (rudymatela) [reply src/WebInterface.hs:269]: That would not work in this case, as this is our internal web interface, `:github-emoji-thing:` is not interpreted as an emoji.

**Comment by rudymatela** [comment]
> The emoji indicators are cute, but as we have more they become less obvious. The blue diamond for promoted feels odd, but I can't think of a good choice for this. Text only also has problems though, because it becomes a bit more difficult to just scan the page with your eyes... What do you think?

It does feel hard to parse in the case I presented.  However in practice we expect to seldom encounter icons other than :yellow_circle: and :x::

* :heavy_check_mark: - this will only appear if the build of the second/third/etc PRs in the train terminates before the build of the first.  This does happen in practice if builds are started almost at the same time, but usually the first PR would be about to finish building.  So this should only be in place for brief window;
* :exclamation: :exclamation: Since the start of the month, rebase failures have only happened 17 times.  That's 0.68 times per day taking into account all 26 projects we track.  If we open the global view and scroll to the page, we expect to see one for a few hours once every couple of days.
* :exclamation: merge rejections due to wrong target branch have happened only once this month!
* :large_blue_diamond: As soon as a PR is promoted, it is deleted from the Hoff state.  So I don't expect this to appear in the UI actually.  I've put the option there in case we end up displaying it (by accident or by future decision).  I've added a comment noting this.

The interface state shown here is a bit of an extreme example illustrating all states:

extreme

In practice we normally see something like:

* Building 
    * ... :yellow_circle: 
    * ... :yellow_circle: 
    * ... :yellow_circle:
* Failed
    * ... :x:
    * ... :x:
    * ... :x:
    * ... :x:

Other icons will be more like exceptions:

* Building 
    * ... :yellow_circle: 
    * ... :yellow_circle: 
    * ... :yellow_circle:
* Failed
    * ... :x:
    * ... :x:
    * ... :exclamation: incorrect base branch
    * ... :x:

... or only shown during a brief window:

* Building 
    * ... :yellow_circle: 
    * ... :heavy_check_mark: 
    * ... :yellow_circle:
* Failed
    * ... :x:
    * ... :x:
    * ... :x:
    * ... :x:

PS: here are the two queries:

```bash
sudo journalctl -u hoff -S2022-08-01 | grep "Posted.*Failed to rebase"
sudo journalctl -u hoff -S2022-08-01 | grep "Posted.*Merge rejected:"
```

File: src/WebInterface.hs (hunk 1 above)
Comment (alex-mckenna) [reply src/WebInterface.hs:269]: Ahh, of course. Silly me

**Comment by alex-mckenna** [comment]
Hmm, okay. That seems reasonable then. Thanks for adding the comment :+1: 

PS: Good to know, you don't need `sudo` for those commands by the way

**Review by alex-mckenna — APPROVED** [review]

**Comment by rudymatela** [comment]
@alex-mckenna Thanks for the review.  :tada:

@OpsBotPrime merge :ship: :shipit: :rocket:

**Comment by OpsBotPrime** [comment]
[CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/12940d9a-8186-4c72-a4e8-f9ed2b5c7a71?pipeline_id=29010f6e-84d1-4653-8d0d-e79ea85d728a) started.
